From f3ee9a6e41734543be4fdd747de303671e69ce94 Mon Sep 17 00:00:00 2001
From: ylevavi <yehuda.levavi@tandemg.com>
Date: Mon, 21 Jun 2021 17:53:53 +0300
Subject: [PATCH] [SW-46440] bmc: install-hls: fix version compare

Fix the way image and current versions are compared so it will work
properly between major, minor and patch numbering.

Change-Id: I09741e37ad68e2bc4fd2f03b42f1741177afc97c
Signed-off-by: ylevavi <yehuda.levavi@tandemg.com>
---

diff --git a/package/openbmc/install-hls.sh b/package/openbmc/install-hls.sh
index 92f18b1..2970326 100644
--- a/package/openbmc/install-hls.sh
+++ b/package/openbmc/install-hls.sh
@@ -142,19 +142,39 @@
 
 function is_version_downgrade
 {
-	local curr_ver=$(echo "${bmc_ver_info}" | grep VERSION_ID | cut -d'=' -f 2 \
-		| sed -e 's/^"//' -e 's/"$//')
-	export curr_major_ver=$(echo ${curr_ver} | cut -d'-' -f 3 | cut -d'.' -f 2)
+	local curr_ver=$(echo "${bmc_ver_info}" | grep VERSION_ID | cut -d'=' -f 2 | sed -e 's/^"//' -e 's/"$//')
+	export curr_ver_num=$(echo ${curr_ver} | cut -d'-' -f 3)
+	export curr_major_ver=$(echo ${curr_ver_num} | cut -d'.' -f 1)
+	export curr_minor_ver=$(echo ${curr_ver_num} | cut -d'.' -f 2)
+	export curr_patch_ver=$(echo ${curr_ver_num} | cut -d'.' -f 3)
 
-	local img_ver=$(tar xfO ${BMC_IMAGE} MANIFEST | grep version | cut -d'=' -f 2)
+	local img_ver_num=$(tar xfO ${BMC_IMAGE} MANIFEST | grep version | cut -d'=' -f 2 | cut -d'-' -f 3)
 	if [ $? -ne 0 ]; then
 		echo Failed to find version manifest
 		exit 2
 	fi
 
-	export img_major_ver=$(echo ${img_ver} | cut -d'.' -f 2)
+	export img_major_ver=$(echo ${img_ver_num} | cut -d'.' -f 1)
+	export img_minor_ver=$(echo ${img_ver_num} | cut -d'.' -f 2)
+	export img_patch_ver=$(echo ${img_ver_num} | cut -d'.' -f 3)
+
+	echo "image ver num: ${img_ver_num}, cur ver num: ${curr_ver_num}"
+	echo "image major: ${img_major_ver}, cur major: ${curr_major_ver}"
+
 	if [ ${img_major_ver} -lt ${curr_major_ver} ]; then
 		return 0
+	elif [ ${img_major_ver} -eq ${curr_major_ver} ]; then
+		if [ ${img_minor_ver} -lt ${curr_minor_ver} ]; then
+			return 0
+		elif [ ${img_minor_ver} -eq ${curr_minor_ver} ]; then
+			if [ ${img_patch_ver} -lt ${curr_patch_ver} ]; then
+				return 0
+			else
+				return 1
+			fi
+		else
+			return 1
+		fi
 	else
 		return 1
 	fi
@@ -212,7 +232,7 @@
 			whole_chip_and_clear_overlay "Downgrading"
 		fi
 
-		echo You are about to downgrade fw version "(${curr_ver} -> ${img_ver})"
+		echo You are about to downgrade fw version "(${curr_ver} -> ${img_ver_num})"
 		echo This means that the flash layout will change
 		read -p "Are you sure you wish to continue? [y/N] " -r
 		if [[ $REPLY =~ ^[Yy]$ ]]; then
